File: python/backtest.py
"""
Backtesting Framework for Cross-Attention Multi-Asset Strategy

Provides:
- CrossAttentionBacktest: Main backtesting class
- Performance metrics computation
- Visualization utilities
"""

import numpy as np
import pandas as pd
from typing import Dict, List, Optional, Tuple
from dataclasses import dataclass
import torch


@dataclass
class BacktestConfig:
    """Configuration for backtesting."""
    initial_capital: float = 100000.0
    transaction_cost: float = 0.001  # 0.1% per trade
    rebalance_freq: int = 24  # Hours between rebalances
    max_position: float = 0.5  # Maximum position per asset
    allow_short: bool = False  # Allow short positions


class CrossAttentionBacktest:
    """
    Backtesting framework for cross-attention portfolio strategy.

    Example:
        backtest = CrossAttentionBacktest(model, config)
        results = backtest.run(X_test, returns_test, timestamps_test)
        metrics = backtest.compute_metrics(results)
    """

    def __init__(
        self,
        model: torch.nn.Module,
        config: BacktestConfig = None
    ):
        self.model = model
        self.config = config or BacktestConfig()

    def run(
        self,
        X: np.ndarray,
        returns: np.ndarray,
        timestamps: pd.DatetimeIndex,
        device: str = 'cpu'
    ) -> pd.DataFrame:
        """
        Run backtest on test data.

        Args:
            X: [n_samples, n_assets, lookback, n_features]
            returns: [n_samples, n_assets] - Actual returns
            timestamps: DatetimeIndex for each sample
            device: 'cpu' or 'cuda'

        Returns:
            DataFrame with backtest results
        """
        self.model.eval()
        self.model = self.model.to(device)

        n_samples = len(X)
        n_assets = X.shape[1]

        capital = self.config.initial_capital
        positions = None  # Will be set on first rebalance

        results = []

        for i in range(0, n_samples, self.config.rebalance_freq):
            # Get model predictions
            with torch.no_grad():
                x_batch = torch.FloatTensor(X[i:i+1]).to(device)
                output = self.model(x_batch)
                weights = output['predictions'].cpu().numpy().flatten()

            # Process weights based on output type
            weights = self._process_weights(weights)

            # Calculate transaction costs
            position_change = np.abs(weights - positions).sum() if positions is not None else weights.sum()
            costs = position_change * self.config.transaction_cost * capital

            # Update positions at start of period to use model predictions immediately
            positions = weights.copy()

            # Simulate trading for each timestep until next rebalance
            period_end = min(i + self.config.rebalance_freq, n_samples)

            for j in range(i, period_end):
                # Calculate portfolio return using current positions
                portfolio_return = np.sum(positions * returns[j])
                capital = capital * (1 + portfolio_return)

                # Deduct costs at first timestep of period
                if j == i:
                    capital -= costs

                results.append({
                    'timestamp': timestamps[j] if j < len(timestamps) else None,
                    'capital': capital,
                    'return': portfolio_return,
                    'positions': positions.copy(),
                    'weights': weights.copy(),
                    'costs': costs if j == i else 0.0
                })

        return pd.DataFrame(results)

    def _process_weights(self, weights: np.ndarray) -> np.ndarray:
        """Process raw model output into portfolio weights."""
        output_type = getattr(self.model, 'output_type', 'regression')

        if output_type == 'portfolio':
            # Already normalized via softmax
            processed = weights
        elif output_type == 'classification':
            # Convert class probabilities to positions
            # Assume: [down, neutral, up] classes
            if len(weights) % 3 != 0:
                raise ValueError(
                    f"Classification output length {len(weights)} is not divisible by 3. "
                    f"Expected output shape: (n_assets * 3,) for [down, neutral, up] classes."
                )
            n_assets = len(weights) // 3
            weights = weights.reshape(n_assets, 3)
            processed = weights[:, 2] - weights[:, 0]  # up - down
        else:
            # Regression: treat as return predictions
            processed = weights

        # Normalize to sum to 1 (long-only) or allow shorts
        if self.config.allow_short:
            # Normalize by sum of absolute values
            total = np.abs(processed).sum() + 1e-8
            processed = processed / total
        else:
            # Long-only: clip negatives and normalize
            processed = np.clip(processed, 0, None)
            total = processed.sum() + 1e-8
            processed = processed / total

        # Apply position limits
        processed = np.clip(processed, -self.config.max_position, self.config.max_position)

        # Renormalize after clipping to maintain full investment
        if self.config.allow_short:
            total = np.abs(processed).sum() + 1e-8
            processed = processed / total
        else:
            total = processed.sum() + 1e-8
            processed = processed / total

        return processed

    def compute_metrics(self, results: pd.DataFrame) -> Dict:
        """
        Compute performance metrics from backtest results.

        Returns:
            Dictionary of performance metrics
        """
        returns = results['return'].values
        capital = results['capital'].values

        # Total Return
        total_return = (capital[-1] / self.config.initial_capital - 1) * 100

        # Annualized metrics (assuming hourly data)
        annual_factor = np.sqrt(365 * 24)

        # Sharpe Ratio
        sharpe = annual_factor * returns.mean() / (returns.std() + 1e-8)

        # Sortino Ratio
        downside_returns = returns[returns < 0]
        downside_std = downside_returns.std() if len(downside_returns) > 0 else 1e-8
        sortino = annual_factor * returns.mean() / (downside_std + 1e-8)

        # Maximum Drawdown
        cumulative = (1 + returns).cumprod()
        running_max = np.maximum.accumulate(cumulative)
        drawdown = (cumulative - running_max) / (running_max + 1e-8)
        max_drawdown = drawdown.min() * 100

        # Calmar Ratio
        annual_return = (capital[-1] / self.config.initial_capital) ** (365 * 24 / len(returns)) - 1
        calmar = annual_return / (abs(max_drawdown / 100) + 1e-8)

        # Win Rate
        win_rate = (returns > 0).mean() * 100

        # Profit Factor
        gross_profit = returns[returns > 0].sum()
        gross_loss = abs(returns[returns < 0].sum())
        profit_factor = gross_profit / (gross_loss + 1e-8)

        # Average Trade
        avg_return = returns.mean() * 100

        # Total Costs
        total_costs = results['costs'].sum()

        return {
            'total_return': total_return,
            'annual_return': annual_return * 100,
            'sharpe_ratio': sharpe,
            'sortino_ratio': sortino,
            'max_drawdown': max_drawdown,
            'calmar_ratio': calmar,
            'win_rate': win_rate,
            'profit_factor': profit_factor,
            'avg_return_pct': avg_return,
            'volatility': returns.std() * annual_factor * 100,
            'total_costs': total_costs,
            'n_trades': len(results) // self.config.rebalance_freq
        }

    def compare_baseline(
        self,
        results: pd.DataFrame,
        returns: np.ndarray
    ) -> Dict:
        """
        Compare strategy against baselines.

        Baselines:
        - Equal Weight: 1/n allocation to each asset
        - Buy & Hold: Hold first asset (typically BTC)
        """
        n_assets = returns.shape[1]

        # Equal Weight baseline
        equal_weight_returns = returns.mean(axis=1)
        equal_weight_cumulative = (1 + equal_weight_returns).cumprod()
        equal_weight_total = (equal_weight_cumulative[-1] - 1) * 100

        # Buy & Hold first asset
        buy_hold_returns = returns[:, 0]
        buy_hold_cumulative = (1 + buy_hold_returns).cumprod()
        buy_hold_total = (buy_hold_cumulative[-1] - 1) * 100

        # Strategy returns
        strategy_total = (results['capital'].iloc[-1] / self.config.initial_capital - 1) * 100

        return {
            'strategy_return': strategy_total,
            'equal_weight_return': equal_weight_total,
            'buy_hold_return': buy_hold_total,
            'outperformance_vs_equal': strategy_total - equal_weight_total,
            'outperformance_vs_buyhold': strategy_total - buy_hold_total
        }


def plot_backtest_results(
    results: pd.DataFrame,
    returns: np.ndarray,
    symbols: List[str],
    save_path: Optional[str] = None
):
    """
    Plot backtest results with equity curve and comparisons.
    """
    try:
        import matplotlib.pyplot as plt
        import matplotlib.dates as mdates
    except ImportError:
        print("matplotlib not available for plotting")
        return

    fig, axes = plt.subplots(3, 1, figsize=(14, 10))

    # 1. Equity Curve
    ax1 = axes[0]
    ax1.plot(results['timestamp'], results['capital'], label='Strategy', linewidth=2)

    # Add baselines
    initial = results['capital'].iloc[0]
    n_assets = returns.shape[1]

    # Equal weight
    equal_weight = initial * (1 + returns.mean(axis=1)).cumprod()
    ax1.plot(results['timestamp'], equal_weight, label='Equal Weight', linestyle='--', alpha=0.7)

    # Buy & hold first asset
    buy_hold = initial * (1 + returns[:, 0]).cumprod()
    ax1.plot(results['timestamp'], buy_hold, label=f'Buy & Hold {symbols[0]}', linestyle=':', alpha=0.7)

    ax1.set_title('Portfolio Equity Curve')
    ax1.set_ylabel('Capital ($)')
    ax1.legend()
    ax1.grid(True, alpha=0.3)

    # 2. Drawdown
    ax2 = axes[1]
    cumulative = results['capital'] / results['capital'].iloc[0]
    running_max = cumulative.cummax()
    drawdown = (cumulative - running_max) / running_max * 100

    ax2.fill_between(results['timestamp'], drawdown, 0, alpha=0.3, color='red')
    ax2.plot(results['timestamp'], drawdown, color='red', linewidth=1)
    ax2.set_title('Drawdown')
    ax2.set_ylabel('Drawdown (%)')
    ax2.grid(True, alpha=0.3)

    # 3. Portfolio Weights over time
    ax3 = axes[2]
    weights_df = pd.DataFrame(
        np.vstack(results['weights'].values),
        columns=symbols,
        index=results['timestamp']
    )

    # Resample to reduce noise
    weights_resampled = weights_df.resample('D').mean()
    weights_resampled.plot.area(ax=ax3, alpha=0.7)

    ax3.set_title('Portfolio Weights Over Time')
    ax3.set_ylabel('Weight')
    ax3.legend(loc='upper left', bbox_to_anchor=(1, 1))
    ax3.grid(True, alpha=0.3)

    plt.tight_layout()

    if save_path:
        plt.savefig(save_path, dpi=150, bbox_inches='tight')
        print(f"Plot saved to {save_path}")
    else:
        plt.show()

    plt.close()


def plot_attention_analysis(
    model: torch.nn.Module,
    X: np.ndarray,
    symbols: List[str],
    save_path: Optional[str] = None
):
    """
    Visualize cross-asset attention patterns.
    """
    try:
        import matplotlib.pyplot as plt
        import seaborn as sns
    except ImportError:
        print("matplotlib/seaborn not available for plotting")
        return

    model.eval()

    with torch.no_grad():
        x = torch.FloatTensor(X)
        output = model(x, return_attention=True)

    attention_weights = output.get('attention_weights')

    if attention_weights is None:
        print("No attention weights available")
        return

    # Get last layer cross-asset attention
    last_layer = list(attention_weights.keys())[-1]
    cross_attn = attention_weights[last_layer].get('cross_asset')

    if cross_attn is None:
        print("No cross-asset attention found")
        return

    # Average over batch
    avg_attn = cross_attn.mean(dim=0).numpy()

    # Plot heatmap
    fig, ax = plt.subplots(figsize=(10, 8))

    sns.heatmap(
        avg_attn,
        xticklabels=symbols,
        yticklabels=symbols,
        annot=True,
        fmt='.2f',
        cmap='Blues',
        ax=ax
    )

    ax.set_title('Cross-Asset Attention Weights\n(Query asset attends to Key asset)')
    ax.set_xlabel('Key (Source Asset)')
    ax.set_ylabel('Query (Target Asset)')

    plt.tight_layout()

    if save_path:
        plt.savefig(save_path, dpi=150, bbox_inches='tight')
        print(f"Attention plot saved to {save_path}")
    else:
        plt.show()

    plt.close()


def print_metrics_report(metrics: Dict, comparison: Dict = None):
    """Print formatted metrics report."""
    print("\n" + "=" * 50)
    print("BACKTEST PERFORMANCE REPORT")
    print("=" * 50)

    print("\n--- Returns ---")
    print(f"Total Return:     {metrics['total_return']:>10.2f}%")
    print(f"Annual Return:    {metrics['annual_return']:>10.2f}%")
    print(f"Avg Return/Trade: {metrics['avg_return_pct']:>10.4f}%")

    print("\n--- Risk Metrics ---")
    print(f"Sharpe Ratio:     {metrics['sharpe_ratio']:>10.2f}")
    print(f"Sortino Ratio:    {metrics['sortino_ratio']:>10.2f}")
    print(f"Calmar Ratio:     {metrics['calmar_ratio']:>10.2f}")
    print(f"Max Drawdown:     {metrics['max_drawdown']:>10.2f}%")
    print(f"Volatility:       {metrics['volatility']:>10.2f}%")

    print("\n--- Trading Stats ---")
    print(f"Win Rate:         {metrics['win_rate']:>10.2f}%")
    print(f"Profit Factor:    {metrics['profit_factor']:>10.2f}")
    print(f"Number of Trades: {metrics['n_trades']:>10d}")
    print(f"Total Costs:      ${metrics['total_costs']:>9.2f}")

    if comparison:
        print("\n--- Baseline Comparison ---")
        print(f"Strategy Return:     {comparison['strategy_return']:>10.2f}%")
        print(f"Equal Weight Return: {comparison['equal_weight_return']:>10.2f}%")
        print(f"Buy & Hold Return:   {comparison['buy_hold_return']:>10.2f}%")
        print(f"vs Equal Weight:     {comparison['outperformance_vs_equal']:>+10.2f}%")
        print(f"vs Buy & Hold:       {comparison['outperformance_vs_buyhold']:>+10.2f}%")

    print("\n" + "=" * 50)


if __name__ == "__main__":
    # Test backtesting framework
    print("Testing backtesting framework...")

    # Create dummy model for testing
    from model import create_model

    model = create_model(
        n_assets=5,
        n_features=6,
        seq_len=48,
        d_model=32,
        output_type='portfolio'
    )

    # Create dummy data
    n_samples = 200
    n_assets = 5
    seq_len = 48
    n_features = 6

    X = np.random.randn(n_samples, n_assets, seq_len, n_features).astype(np.float32)
    returns = np.random.randn(n_samples, n_assets) * 0.01  # Small returns
    timestamps = pd.date_range('2024-01-01', periods=n_samples, freq='H')

    symbols = ['BTC', 'ETH', 'SOL', 'AVAX', 'DOT']

    # Run backtest
    config = BacktestConfig(
        initial_capital=100000,
        rebalance_freq=24
    )

    backtest = CrossAttentionBacktest(model, config)
    results = backtest.run(X, returns, timestamps)

    print(f"\nBacktest completed: {len(results)} timesteps")

    # Compute metrics
    metrics = backtest.compute_metrics(results)
    comparison = backtest.compare_baseline(results, returns)

    print_metrics_report(metrics, comparison)

    print("\nAll backtest tests passed!")

File: rust/src/strategy/backtest.rs
//! Backtesting engine for Cross-Attention Multi-Asset Trading
//!
//! Evaluates trading strategies with realistic constraints.

use serde::{Deserialize, Serialize};

/// Backtest configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BacktestConfig {
    /// Initial capital
    pub initial_capital: f64,
    /// Transaction cost (as fraction)
    pub transaction_cost: f64,
    /// Rebalance frequency (in time steps)
    pub rebalance_freq: usize,
    /// Maximum position size per asset
    pub max_position: f64,
    /// Allow short positions
    pub allow_short: bool,
    /// Slippage (as fraction)
    pub slippage: f64,
}

impl Default for BacktestConfig {
    fn default() -> Self {
        Self {
            initial_capital: 100_000.0,
            transaction_cost: 0.001, // 10 bps
            rebalance_freq: 24,      // Daily for hourly data
            max_position: 0.5,
            allow_short: false,
            slippage: 0.0005, // 5 bps
        }
    }
}

/// Single step result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StepResult {
    /// Timestamp
    pub timestamp: i64,
    /// Portfolio value
    pub portfolio_value: f64,
    /// Portfolio weights
    pub weights: Vec<f64>,
    /// Returns for this step
    pub step_return: f64,
    /// Transaction costs incurred
    pub transaction_costs: f64,
}

/// Backtest result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BacktestResult {
    /// Step-by-step results
    pub steps: Vec<StepResult>,
    /// Performance metrics
    pub metrics: PerformanceMetrics,
    /// Asset symbols
    pub symbols: Vec<String>,
}

/// Performance metrics
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PerformanceMetrics {
    /// Total return
    pub total_return: f64,
    /// Annualized return
    pub annualized_return: f64,
    /// Sharpe ratio (annualized)
    pub sharpe_ratio: f64,
    /// Sortino ratio (annualized)
    pub sortino_ratio: f64,
    /// Calmar ratio
    pub calmar_ratio: f64,
    /// Maximum drawdown
    pub max_drawdown: f64,
    /// Volatility (annualized)
    pub volatility: f64,
    /// Win rate
    pub win_rate: f64,
    /// Total transaction costs
    pub total_transaction_costs: f64,
    /// Number of trades
    pub n_trades: usize,
}

impl Default for PerformanceMetrics {
    fn default() -> Self {
        Self {
            total_return: 0.0,
            annualized_return: 0.0,
            sharpe_ratio: 0.0,
            sortino_ratio: 0.0,
            calmar_ratio: 0.0,
            max_drawdown: 0.0,
            volatility: 0.0,
            win_rate: 0.0,
            total_transaction_costs: 0.0,
            n_trades: 0,
        }
    }
}

/// Backtesting engine
pub struct Backtest {
    config: BacktestConfig,
}

impl Backtest {
    /// Create a new backtest engine
    pub fn new(config: BacktestConfig) -> Self {
        Self { config }
    }

    /// Run backtest
    ///
    /// # Arguments
    /// * `weights` - Model-predicted weights [n_steps, n_assets]
    /// * `returns` - Actual returns [n_steps, n_assets]
    /// * `symbols` - Asset symbols
    /// * `timestamps` - Timestamps for each step
    ///
    /// # Returns
    /// Backtest result with step-by-step data and metrics
    pub fn run(
        &self,
        weights: &[Vec<f64>],
        returns: &[Vec<f64>],
        symbols: &[String],
        timestamps: &[i64],
    ) -> BacktestResult {
        assert_eq!(weights.len(), returns.len(), "Weights and returns must have same length");
        assert!(!weights.is_empty(), "Must have at least one step");

        let n_steps = weights.len();
        let n_assets = symbols.len();

        let mut steps = Vec::with_capacity(n_steps);
        let mut portfolio_value = self.config.initial_capital;
        let mut current_weights = vec![1.0 / n_assets as f64; n_assets]; // Equal weight start
        let mut total_costs = 0.0;
        let mut n_trades = 0;

        for i in 0..n_steps {
            let target_weights = &weights[i];
            let asset_returns = &returns[i];

            // Check if rebalancing
            let should_rebalance = i % self.config.rebalance_freq == 0;

            // Calculate transaction costs if rebalancing
            let transaction_cost = if should_rebalance {
                let cost = self.calculate_transaction_cost(&current_weights, target_weights, portfolio_value);
                total_costs += cost;
                n_trades += 1;
                cost
            } else {
                0.0
            };

            // Apply transaction costs
            portfolio_value -= transaction_cost;

            // Update weights if rebalancing
            if should_rebalance {
                current_weights = self.apply_constraints(target_weights);
            }

            // Calculate portfolio return
            let portfolio_return: f64 = current_weights
                .iter()
                .zip(asset_returns.iter())
                .map(|(w, r)| w * r)
                .sum();

            // Apply slippage if rebalancing
            let slippage_cost = if should_rebalance {
                portfolio_value * self.config.slippage
            } else {
                0.0
            };
            portfolio_value -= slippage_cost;

            // Update portfolio value
            portfolio_value *= 1.0 + portfolio_return;

            // Update weights based on returns (drift)
            if !should_rebalance {
                let total: f64 = current_weights
                    .iter()
                    .zip(asset_returns.iter())
                    .map(|(w, r)| w * (1.0 + r))
                    .sum();

                if total > 0.0 {
                    for (w, r) in current_weights.iter_mut().zip(asset_returns.iter()) {
                        *w = *w * (1.0 + r) / total;
                    }
                }
            }

            steps.push(StepResult {
                timestamp: timestamps.get(i).copied().unwrap_or(i as i64),
                portfolio_value,
                weights: current_weights.clone(),
                step_return: portfolio_return,
                transaction_costs: transaction_cost + slippage_cost,
            });
        }

        // Calculate metrics
        let metrics = self.calculate_metrics(&steps, total_costs, n_trades);

        BacktestResult {
            steps,
            metrics,
            symbols: symbols.to_vec(),
        }
    }

    /// Apply position constraints
    fn apply_constraints(&self, weights: &[f64]) -> Vec<f64> {
        let mut result: Vec<f64> = weights
            .iter()
            .map(|w| {
                if self.config.allow_short {
                    w.clamp(-self.config.max_position, self.config.max_position)
                } else {
                    w.clamp(0.0, self.config.max_position)
                }
            })
            .collect();

        // Normalize to sum to 1 (for long-only)
        if !self.config.allow_short {
            let total: f64 = result.iter().sum();
            if total > 0.0 {
                for w in &mut result {
                    *w /= total;
                }
            }
        }

        result
    }

    /// Calculate transaction cost
    fn calculate_transaction_cost(
        &self,
        current: &[f64],
        target: &[f64],
        portfolio_value: f64,
    ) -> f64 {
        let turnover: f64 = current
            .iter()
            .zip(target.iter())
            .map(|(c, t)| (t - c).abs())
            .sum();

        portfolio_value * turnover * self.config.transaction_cost
    }

    /// Calculate performance metrics
    fn calculate_metrics(
        &self,
        steps: &[StepResult],
        total_costs: f64,
        n_trades: usize,
    ) -> PerformanceMetrics {
        if steps.is_empty() {
            return PerformanceMetrics::default();
        }

        let n = steps.len();
        let final_value = steps.last().unwrap().portfolio_value;
        let initial_value = self.config.initial_capital;

        // Total return
        let total_return = (final_value - initial_value) / initial_value;

        // Returns series
        let returns: Vec<f64> = steps.iter().map(|s| s.step_return).collect();

        // Annualized return (assuming hourly data, 24*365 periods per year)
        let periods_per_year = 24.0 * 365.0;
        let annualized_return = (1.0 + total_return).powf(periods_per_year / n as f64) - 1.0;

        // Volatility
        let mean_return: f64 = returns.iter().sum::<f64>() / n as f64;
        let variance: f64 = returns.iter().map(|r| (r - mean_return).powi(2)).sum::<f64>() / n as f64;
        let std_dev = variance.sqrt();
        let volatility = std_dev * (periods_per_year as f64).sqrt();

        // Sharpe ratio (assuming 0% risk-free rate)
        let sharpe_ratio = if volatility > 0.0 {
            annualized_return / volatility
        } else {
            0.0
        };

        // Sortino ratio (downside deviation)
        let downside_returns: Vec<f64> = returns.iter().filter(|&&r| r < 0.0).copied().collect();
        let downside_variance = if !downside_returns.is_empty() {
            downside_returns.iter().map(|r| r.powi(2)).sum::<f64>() / downside_returns.len() as f64
        } else {
            0.0
        };
        let downside_deviation = downside_variance.sqrt() * (periods_per_year as f64).sqrt();
        let sortino_ratio = if downside_deviation > 0.0 {
            annualized_return / downside_deviation
        } else {
            0.0
        };

        // Maximum drawdown
        let mut max_drawdown = 0.0;
        let mut peak = initial_value;
        for step in steps {
            if step.portfolio_value > peak {
                peak = step.portfolio_value;
            }
            let drawdown = (peak - step.portfolio_value) / peak;
            if drawdown > max_drawdown {
                max_drawdown = drawdown;
            }
        }

        // Calmar ratio
        let calmar_ratio = if max_drawdown > 0.0 {
            annualized_return / max_drawdown
        } else {
            0.0
        };

        // Win rate
        let positive_returns = returns.iter().filter(|&&r| r > 0.0).count();
        let win_rate = positive_returns as f64 / n as f64;

        PerformanceMetrics {
            total_return,
            annualized_return,
            sharpe_ratio,
            sortino_ratio,
            calmar_ratio,
            max_drawdown,
            volatility,
            win_rate,
            total_transaction_costs: total_costs,
            n_trades,
        }
    }

    /// Compare with equal-weight baseline
    pub fn compare_baseline(
        &self,
        returns: &[Vec<f64>],
        symbols: &[String],
        timestamps: &[i64],
    ) -> BacktestResult {
        let n_assets = symbols.len();
        let equal_weights: Vec<Vec<f64>> = returns
            .iter()
            .map(|_| vec![1.0 / n_assets as f64; n_assets])
            .collect();

        self.run(&equal_weights, returns, symbols, timestamps)
    }
}

impl Default for Backtest {
    fn default() -> Self {
        Self::new(BacktestConfig::default())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_test_data(n_steps: usize, n_assets: usize) -> (Vec<Vec<f64>>, Vec<Vec<f64>>, Vec<String>, Vec<i64>) {
        let weights: Vec<Vec<f64>> = (0..n_steps)
            .map(|_| {
                let mut w = vec![1.0 / n_assets as f64; n_assets];
                w[0] += 0.1;
                w[1] -= 0.1 / (n_assets - 1) as f64;
                let total: f64 = w.iter().sum();
                w.iter().map(|x| x / total).collect()
            })
            .collect();

        let returns: Vec<Vec<f64>> = (0..n_steps)
            .map(|i| {
                (0..n_assets)
                    .map(|j| 0.001 * ((i + j) as f64 * 0.1).sin())
                    .collect()
            })
            .collect();

        let symbols: Vec<String> = (0..n_assets).map(|i| format!("ASSET{}", i)).collect();
        let timestamps: Vec<i64> = (0..n_steps).map(|i| i as i64 * 3600000).collect();

        (weights, returns, symbols, timestamps)
    }

    #[test]
    fn test_backtest_creation() {
        let config = BacktestConfig::default();
        let backtest = Backtest::new(config);

        assert_eq!(backtest.config.initial_capital, 100_000.0);
    }

    #[test]
    fn test_backtest_run() {
        let (weights, returns, symbols, timestamps) = create_test_data(100, 5);
        let backtest = Backtest::default();

        let result = backtest.run(&weights, &returns, &symbols, &timestamps);

        assert_eq!(result.steps.len(), 100);
        assert!(result.metrics.total_return.is_finite());
    }

    #[test]
    fn test_backtest_metrics() {
        let (weights, returns, symbols, timestamps) = create_test_data(1000, 5);
        let backtest = Backtest::default();

        let result = backtest.run(&weights, &returns, &symbols, &timestamps);
        let metrics = &result.metrics;

        // Verify metrics are calculated
        assert!(metrics.sharpe_ratio.is_finite());
        assert!(metrics.max_drawdown >= 0.0);
        assert!(metrics.max_drawdown <= 1.0);
        assert!(metrics.win_rate >= 0.0);
        assert!(metrics.win_rate <= 1.0);
    }

    #[test]
    fn test_baseline_comparison() {
        let (_, returns, symbols, timestamps) = create_test_data(100, 5);
        let backtest = Backtest::default();

        let result = backtest.compare_baseline(&returns, &symbols, &timestamps);

        assert_eq!(result.steps.len(), 100);
        // Equal weights should sum to 1
        let weight_sum: f64 = result.steps[0].weights.iter().sum();
        assert!((weight_sum - 1.0).abs() < 1e-6);
    }

    #[test]
    fn test_transaction_costs() {
        let (weights, returns, symbols, timestamps) = create_test_data(100, 5);
        let config = BacktestConfig {
            transaction_cost: 0.01, // 1% to make it significant
            ..Default::default()
        };
        let backtest = Backtest::new(config);

        let result = backtest.run(&weights, &returns, &symbols, &timestamps);

        assert!(result.metrics.total_transaction_costs > 0.0);
        assert!(result.metrics.n_trades > 0);
    }

    #[test]
    fn test_position_constraints() {
        let config = BacktestConfig {
            max_position: 0.3,
            ..Default::default()
        };
        let backtest = Backtest::new(config);

        let weights = vec![0.8, 0.1, 0.05, 0.05];
        let constrained = backtest.apply_constraints(&weights);

        // After capping at 0.3: [0.3, 0.1, 0.05, 0.05] sum = 0.5
        // After normalization: [0.6, 0.2, 0.1, 0.1] sum = 1.0
        // The constraint is applied before normalization
        let sum: f64 = constrained.iter().sum();
        assert!((sum - 1.0).abs() < 1e-6, "Weights should sum to 1");

        // Check that no weight is negative
        for w in &constrained {
            assert!(*w >= 0.0, "No negative weights");
        }
    }
}
